core/client/hammer/hammer.go

```diff
@@ -76,125 +93,159 @@ func New(ctx context.Context, dial DialFunc, callOptions CallOptions,
-	h.printStats(1*time.Second, outputs)
-}
-
-// launchWorkers gradually adds workers to the worker pool over ramp up to workers.
-func (h *Hammer) launchWorkers(ctx context.Context, jobs <-chan int, workers int, ramp time.Duration) <-chan error {
-	outputs := make(chan error)
-
-	go func() {
-		rampTicker := time.NewTicker(ramp / time.Duration(workers))
-		var wg sync.WaitGroup
-		for h.workers < workers {
-			h.workers++
-			wg.Add(1)
-			go func() {
-				defer wg.Done()
-				h.worker(ctx, jobs, outputs)
-			}()
-			<-rampTicker.C
-		}
-		rampTicker.Stop()
+// workers should roughly be (goal QPS / timeout seconds)
+func (h *Hammer) Run(ctx context.Context, numWorkers int, c Config) error {
+	workers, err := h.newWorkers(numWorkers)
+	if err != nil {
+		return err
+	}
 
-		wg.Wait()
-		close(outputs)
-	}()
+	// Batch Write users
+	log.Printf("Batch Write")
+	requests := genRequests(ctx, c.BatchWriteQPS, c.BatchWriteSize, c.BatchWriteCount, c.Duration)
+	handlers := make([]ReqHandler, 0, len(workers))
+	for _, w := range workers {
+		handlers = append(handlers, w.writeOp)
+	}
+	generateReport(ctx, requests, handlers)
+
+	// Write users
+	log.Printf("User Write")
+	requests = genRequests(ctx, c.WriteQPS, 1, c.WriteCount, c.Duration)
+	handlers = make([]ReqHandler, 0, len(workers))
+	for _, w := range workers {
+		handlers = append(handlers, w.writeOp)
+	}
+	generateReport(ctx, requests, handlers)
+
+	// Read users
+	log.Printf("User Read")
+	requests = genRequests(ctx, c.ReadQPS, 1, c.ReadCount, c.Duration)
+	handlers = make([]ReqHandler, 0, len(workers))
+	for _, w := range workers {
+		handlers = append(handlers, w.readOp)
+	}
+	generateReport(ctx, requests, handlers)
+
+	// History
+	log.Printf("User Audit History")
+	requests = genRequests(ctx, c.HistoryQPS, c.HistoryBatchSize, c.HistoryCount, c.Duration)
+	handlers = make([]ReqHandler, 0, len(workers))
+	for _, w := range workers {
+		handlers = append(handlers, w.historyOp)
+	}
+	generateReport(ctx, requests, handlers)
 
-	return outputs
+	return nil
 }
 
-// gen produces a channel of integers from 0 up to count-1 and then closes.
-func gen(ctx context.Context, count int) <-chan int {
-	jobs := make(chan int)
+func genRequests(ctx context.Context, qps, batch, count int, duration time.Duration) <-chan request {
+	inflightReqs := make(chan request)
 	go func() {
-		defer close(jobs)
+		cctx, cancel := context.WithTimeout(ctx, duration)
+		defer cancel()
+		defer close(inflightReqs)
+
+		rateLimiter := rate.NewLimiter(rate.Limit(qps), qps)
 		for i := 0; i < count; i++ {
-			select {
-			case <-ctx.Done():
+			userIDs := make([]string, 0, batch)
+			for j := 0; j < batch; j++ {
+				userIDs = append(userIDs, fmt.Sprintf("user_%v", i*batch+j))
+			}
+			if err := rateLimiter.Wait(cctx); err != nil {
 				return
-			case jobs <- i:
+			}
+			inflightReqs <- request{
+				UserIDs:   userIDs,
+				BatchSize: batch,
 			}
 		}
 	}()
-	return jobs
+	return inflightReqs
 }
 
```

Comment:
I like that there's no concurrency-related stuff from here onwards in the file, just a bunch of linear functions -- it makes it much easier to think about the concurrency when it's concentrated in a smaller number of places.

Reply:
👍